### ISSUE TITLE
Log_softmax in KL Divergence Loss

### ISSUE THREAD

**Issue by alessiomora** [jinkyu032/FedMLB#3]
I think that your solution is very elegant and very novel for FL.
I would like to ask you two questions:
1. In the code, I see that -- when comparing the hybrid path outputs and the main path output -- you use a ```log_softmax``` for the main path output while a regular ```softmax``` activation for the hybrid path. However, it is not explicitly said or explained in the paper. Why do you use the ```log_softmax```? Can you give an intuition about that?
```
# In FedMLB.py, lines 12-17
def KD(input_p, input_q, T=1):
    kl_loss = nn.KLDivLoss(reduction="batchmean")
    p = F.softmax(input_p/T, dim = 1)
    q = F.log_softmax(input_q/T, dim = 1)
    result = kl_loss(q,p)
    return result   
```
2. Have you tried with both regular softmax activations? Does this worsen the performance of FedMLB?

Thank you in advance, and again, very nice work!

**Comment by jinkyu032** [comment]
Hi, allessiomora.
1.That part is implementation of following part
![image](https://user-images.githubusercontent.com/86775100/207111444-be530b59-51dd-4253-8885-c567f433f54e.png)
which is Kullback-Leibler divergence between hybrid pathway's output and main pathway's output.(equation 7 on page 4 on paper).
It use KLDivLoss function from pytorch library(https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html), and if we look the definition of pointwise KL-divergence,
![image](https://user-images.githubusercontent.com/86775100/207113558-000af5ea-d25b-497a-b4f4-9334d64be4f5.png)
we can find that 
-Related to $y_{pred}$, we need just $log (y_{pred})$
-Related to $y_{true}$, we need both $log (y_{true})$ and $y_{true}$.
So, there's two option to give value related to $y_{true}$ as input to KLDivLoss function and it is controlled by log_target boolean.
![image](https://user-images.githubusercontent.com/86775100/207117667-e0d82484-3d9e-4c39-a1ad-9946b98ced7f.png)

Our implementation is the case with log_target=False, which is default case.
So we use log_softmax for our prediction q
while using softmax to the target p.
(If we use log_target=True option, we can use log_softmax for both)

Examples on the above link may be helpful for more understanding.
2.  I have not tried both regular softmax activation.
If we use both regular softmax activation with "log_target=False"(default) option then it will result in 
$y_{true} * (log (y_{true}) - (y_{pred}))$, which is somewhat different from KL divergence.

Thanks for your attention on our work!

**Comment by alessiomora** [comment]
Thank you very much for the detailed answer. I was not recalling that in pytorch, KLDivLoss expects a log_softmax input. Said that, my point 2 is pointless now.
